api/routes.go

```diff
@@ -24,75 +24,84 @@ var (
 	ErrInvalidAPIKey = errors.New("invalid api key")
 )
 
+type (
+	// HandlerWithUser is a wrapper for httprouter.Handle which also includes
+	// a user parameter. This allows us to fetch the user making the request
+	// just once, during validation.
+	HandlerWithUser func(*database.User, http.ResponseWriter, *http.Request, httprouter.Params)
```

Comment:
I actually kinda like it. Because without it we need to try and grab the user from the context in every handler again and if it doesn't exist we need to return the unauthorized status code. Unless we just let the code run into `nil` pointer dereferences which is also not great. So we are basically back to the same verbose code as before with returning.

But maybe we can do both. So for handlers which absolutely must authenticate we use this and the ones that can be authenticated check the context.

Reply:
Well you'd do something like `userFromContext` and drop a build.Critical alongside an unauthorized response indeed . I don't really mind either way I just think it's weird that a lot of our handlers have `(_ *database.User, w http.ResponseWriter)`

Reply:
feel free to resolve this though it was just a remark